### PR TITLE
Generalize small dominators optimization

### PR DESCRIPTION
**Pull request by tmiasko**
* Use small dominators optimization from 640ede7b0a1840415cb6ec881c2210302bfeba18 more generally.
* Merge `DefLocation` and `LocationExtended` since they serve the same purpose.

### CODE REVIEW

File: compiler/rustc_data_structures/src/graph/dominators/mod.rs [hunk 1]
```diff
@@ -26,7 +26,42 @@ rustc_index::newtype_index! {
...
-pub fn dominators<G: ControlFlowGraph>(graph: &G) -> Dominators<G::Node> {
+#[derive(Clone, Debug)]
+pub struct Dominators<Node: Idx> {
+    kind: Kind<Node>,
+}
+
+#[derive(Clone, Debug)]
+enum Kind<Node: Idx> {
+    /// A representation optimized for a small path graphs.
+    Path(usize),
```
Comment (cjgillot) [inline compiler/rustc_data_structures/src/graph/dominators/mod.rs:37]: Could you say what is the `usize` in the doc?

File: compiler/rustc_mir_transform/src/ssa.rs [hunk 2]
```diff
@@ -187,15 +187,9 @@ impl SsaLocals {
     }
 }
 
-#[derive(Copy, Clone, Debug, PartialEq, Eq)]
-enum LocationExtended {
-    Plain(Location),
-    Arg,
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/ssa.rs:230]: https://github.com/rust-lang/rust/pull/113915 adds a variant to this enum to represent an assignment along an edge. That's useful for call and yield terminators. Can this be modelled with `DefLocation`, or will I have to revert this?

Comment (cjgillot) [reply compiler/rustc_mir_transform/src/ssa.rs:230]: Thinking a bit more about this, reusing `DefLocation` may be the right solution. `rustc_codegen_ssa` only relies on it to compute domination. The difference is the semantics we give to call and yield terminators.

#113915 puts the assignment along the edge. `rustc_codegen_ssa` puts the assignment *before* the edge. They are equivalent iff using the local's value on the unwind edge is UB. Equivalently, that would mean assigning `undef` on the unwind edge.

I guess this is a question for @rust-lang/opsem?

Comment (RalfJung) [reply compiler/rustc_mir_transform/src/ssa.rs:230]: So concretely the question is, what is guaranteed about the contents of the return place when the function unwinds?

That's a good question I hadn't considered yet. Miri will de-initialize its content before calling the function, but anything the callee writes to the return place before unwinding is guaranteed to remain there.

Comment (cjgillot) [reply compiler/rustc_mir_transform/src/ssa.rs:230]: I agree. The better solution is to fix the codegen implementation and then share the result.

**Comment by cjgillot** [comment]
I'll fix the implementation for both MIR and codegen in https://github.com/rust-lang/rust/pull/113915.
@bors r+

File: compiler/rustc_mir_transform/src/ssa.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir_transform/src/ssa.rs:230]: On the semantic question, I realized what I said above was wrong, the caller-provided return place will always be fully uninitialized in Miri upon unwinding. I opened https://github.com/rust-lang/unsafe-code-guidelines/issues/468 to potentially discuss this.

I don't understand what your conclusion means for the semantics, but this should give optimizations and codegen maximal freedom.

**Pull request merged**